src/game/notation.rs

```diff
@@ -148,37 +150,67 @@ fn process_suffix(notation: &str) -> (bool, bool, bool, &str) {
+
     // todo: validate notation only contains low-value utf-8 characters
-    let (checkmate, check, enpassant, ex_suffix) = process_suffix(&notation);
-    let (king_side_castle, queen_side_castle, ex_castling) = process_castling(ex_suffix);
-    let (promoted_to_piece_type, ex_promotion) = process_promotion(ex_castling);
-    let (rank, file, ex_destination) = process_coordinates(ex_promotion);
-    let (capture, ex_capture) = process_capture(ex_destination);
-    let (piece_type, ex_piece_type) = process_piece_type(ex_capture);
-    let (source_rank, source_file, ex_source_coordinates) = process_coordinates(ex_piece_type);
-    match ex_source_coordinates {
-        "" => {
-            let parsed = Notation {
-                text: notation,
-                rank: rank,
-                file: file,
-                piece_type: piece_type,
-                capture: capture,
-                from_rank: source_rank,
-                from_file: source_file,
-                check: check,
-                checkmate: checkmate,
-                enpassant: enpassant,
-                queen_side_castle: queen_side_castle,
-                king_side_castle: king_side_castle,
-                promoted_to_piece_type: promoted_to_piece_type
-            };
-            parsed.validate_parsing();
-            parsed
-        },
-        _ => { panic!("Invalid notation: {}", notation); },
+
+    let mut parsed = Notation {
+        text: "".to_string(), // assigned after parsing
+        rank: None,
+        file: None,
+        piece_type: None,
+        capture: false,
+        from_rank: None,
+        from_file: None,
+        check: false,
+        checkmate: false,
+        enpassant: false,
+        queen_side_castle: false,
+        king_side_castle: false,
+        promoted_to_piece_type: None,
+        end_of_game: None
+    };
+    let (king_side_castle, queen_side_castle) = process_castling(&notation);
+    let end_of_game = process_end_of_game(&notation);
+    if king_side_castle || queen_side_castle || end_of_game.is_some() {
+        parsed.king_side_castle = king_side_castle;
+        parsed.queen_side_castle = queen_side_castle;
+        parsed.end_of_game = end_of_game;
+    } else {
+        let (checkmate, check, enpassant, ex_suffix) = process_suffix(&notation);
+        let (promoted_to_piece_type, ex_promotion) = process_promotion(ex_suffix);
+        let (rank, file, ex_destination) = process_coordinates(ex_promotion);
+        let (capture, ex_capture) = process_capture(ex_destination);
+        let (piece_type, ex_piece_type) = process_piece_type(ex_capture);
+        let (source_rank, source_file, ex_source_coordinates) = process_coordinates(ex_piece_type);
+        match ex_source_coordinates {
+            "" => {
+                parsed.rank = rank;
+                parsed.file = file;
+                parsed.piece_type = piece_type;
+                parsed.capture = capture;
+                parsed.from_rank = source_rank;
```

Comment:
Please review naming conversions in these two lines for consistency

Reply:
good point. In fact, while I'm at it I think I'll change `parsed.rank` and `parsed.file` to `parsed.to_rank` and `parsed.to_file` respectively. Might as well be clear about their meaning

Reply:
there we go - should be better now

Reply:
By the way, in a scenario like this, it would be useful for you to use the "Viewed" checkboxes as you view each file. This would mean that github shows you which of the "Viewed" files have changed when I push an update.

these ones - ![Screen Shot 2019-11-22 at 10 42 48](https://user-images.githubusercontent.com/910448/69419784-1837b600-0d15-11ea-9168-808940b361a5.png)